sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```diff
@@ -4701,4 +4987,425 @@ static UUID randomUuid(long msb, long lsb) {
+            return EMPTY_REGION_LIST;
+        }
+
+        List<URI> endpoints = getApplicableEndPoints(operationType, options);
+
+        HashSet<String> normalizedExcludedRegions = new HashSet<>();
+        if (options.getExcludeRegions() != null) {
+            options
+                .getExcludeRegions()
+                .forEach(r -> normalizedExcludedRegions.add(r.toLowerCase(Locale.ROOT)));
+        }
+
+        List<String> orderedRegionsForSpeculation = new ArrayList<>();
+        endpoints.forEach(uri -> {
+            String regionName = this.globalEndpointManager.getRegionName(uri, operationType);
+            if (!normalizedExcludedRegions.contains(regionName.toLowerCase(Locale.ROOT))) {
+                orderedRegionsForSpeculation.add(regionName);
+            }
+        });
+
+        return orderedRegionsForSpeculation;
+    }
+
+    @FunctionalInterface
+    private interface DocumentPointOperation {
+        Mono<ResourceResponse<Document>> apply(RequestOptions requestOptions, CosmosEndToEndOperationLatencyPolicyConfig endToEndOperationLatencyPolicyConfig, DiagnosticsClientContext clientContextOverride);
+    }
+
+    private static class NonTransientPointOperationResult {
+        private final ResourceResponse<Document> response;
+        private final CosmosException exception;
+
+        public NonTransientPointOperationResult(CosmosException exception) {
+            checkNotNull(exception, "Argument 'exception' must not be null.");
+            this.exception = exception;
+            this.response = null;
+        }
+
+        public NonTransientPointOperationResult(ResourceResponse<Document> response) {
+            checkNotNull(response, "Argument 'response' must not be null.");
+            this.exception = null;
+            this.response = response;
+        }
+
+        public boolean isError() {
+            return this.exception != null;
+        }
+
+        public CosmosException getException() {
+            return this.exception;
+        }
+
+        public ResourceResponse<Document> getResponse() {
+            return this.response;
+        }
+    }
+
+    private static class ScopedDiagnosticsFactory implements DiagnosticsClientContext {
+
+        private AtomicBoolean isMerged = new AtomicBoolean(false);
+        private final DiagnosticsClientContext inner;
+        private final ConcurrentLinkedQueue<CosmosDiagnostics> createdDiagnostics;
+
+        public ScopedDiagnosticsFactory(DiagnosticsClientContext inner) {
+            checkNotNull(inner, "Argument 'inner' must not be null.");
+            this.inner = inner;
+            this.createdDiagnostics = new ConcurrentLinkedQueue<>();
+        }
+
+        @Override
+        public DiagnosticsClientConfig getConfig() {
+            return inner.getConfig();
+        }
+
+        @Override
+        public CosmosDiagnostics createDiagnostics() {
+            assert !this.isMerged.get() : "Not allowed to create new diagnostics after merging.";
```

Comment:
should we throw an exception here or just return null and log the error?

could there be a such race condition happens that:

- we get response from one region,
-  diagnostics merged
- even though it tried to cancel the request to the other regions, but still we reached here for the other regions

Reply:
Yes - I added the assert to validate the assumptions during testing and forgot to revert it.

Reply:
Fixed